### PR TITLE
Convert SaveJobRun to explicit SQL

### PR DESCRIPTION
**Pull request by samsondav**

### CODE REVIEW

File: core/store/orm/orm.go [hunk 1]
```diff
@@ -291,41 +295,123 @@ func (orm *ORM) convenientTransaction(callback func(*gorm.DB) error) error {
...
 		if result.RowsAffected == 0 {
 			return ErrOptimisticUpdateConflict
 		}
-		err := dbtx.Unscoped().Save(&run.Result).Error
-		if err != nil {
-			return err
+
+		uRunResultStrs := []string{}
+		uRunResultArgs := []interface{}{}
+		if run.Result.ID > 0 {
```
Comment (connorwstein) [inline core/store/orm/orm.go:327]: is run.Result guaranteed to be preloaded here?

File: core/store/orm/orm.go [hunk 2]
```diff
@@ -291,41 +295,123 @@ func (orm *ORM) convenientTransaction(callback func(*gorm.DB) error) error {
...
 		}
-		if len(run.TaskRuns) > 0 {
-			err = dbtx.Unscoped().Save(&run.TaskRuns).Error
+
+		uTaskRunStrs := []string{}
+		uTaskRunArgs := []interface{}{}
+		for _, tr := range run.TaskRuns {
+			uTaskRunStrs = append(uTaskRunStrs, "(?::uuid, ?::run_status)")
+			uTaskRunArgs = append(uTaskRunArgs, tr.ID, tr.Status)
+			if tr.Result.ID > 0 {
```
Comment (connorwstein) [inline core/store/orm/orm.go:354]: same here - are guaranteed to have the Result preloaded? Might be good to have a defensive check anyways

File: core/store/orm/orm.go [hunk 3]
```diff
@@ -291,41 +295,123 @@ func (orm *ORM) convenientTransaction(callback func(*gorm.DB) error) error {
...
+			err := dbtx.Exec(stmt, uTaskRunArgs...).Error
 			if err != nil {
-				return err
+				return errors.Wrap(err, "failed to update task runs")
 			}
 		}
-		return nil
+
+		if len(uRunResultStrs) > 0 {
+			updateTaskRunResultsSQL := `
```
Comment (connorwstein) [inline core/store/orm/orm.go:394]: this is actually a mix of job run and task run updates right? I think we need a better var name and a comment about this polymorphism

File: core/store/orm/orm.go (hunk 1 above)
Comment (samsondav) [reply core/store/orm/orm.go:327]: I don't know, but if it isn't then we will return an error on line 332.

File: core/store/orm/orm.go (hunk 2 above)
Comment (samsondav) [reply core/store/orm/orm.go:354]: The defensive check is on line 358 below

Comment (samsondav) [reply core/store/orm/orm.go:357]: I added tests for both of these cases

**Review by connorwstein — APPROVED** [review]

**Pull request merged**